### PR TITLE
Limit the concurrency of password hashing to the number of CPU cores available

### LINKED ISSUE #28477: The concurrency of hashing leads to increased memory usage and CPU throttling
### Description

With the new Argon2 password hashing provider parallel hashing not only leads to CPU throttling, but also an increase memory usage, even an OOM with the default memory settings of Keycloak.

The parallel hashing should be limited to the number of CPU cores available.

### Motivation

The Argon2 password hashing is CPU bound, so it doesn't make sense to hash more values concurrently than there are cores on the machine.
When we run more, this only leads to an increased memory usage and to throttling of the process in containerized environments when a CPU limit is imposed. The throttling would have a negative impact on other concurrent non-hashing activities of Keycloak.

### PR DESCRIPTION
**Pull request by ahus1**
Closes #28477

### CODE REVIEW

File: crypto/default/src/main/java/org/keycloak/crypto/hash/Argon2PasswordHashProviderFactory.java [hunk 1]
```diff
@@ -23,6 +24,13 @@ public class Argon2PasswordHashProviderFactory implements PasswordHashProviderFa
     public static final String ITERATIONS_KEY = "iterations";
     public static final String PARALLELISM_KEY = "parallelism";
 
+    /**
+     * The Argon2 password hashing is CPU bound, so it doesn't make sense to hash more values concurrently than there are cores on the machine.
+     * When we run more, this only leads to an increased memory usage and to throttling of the process in containerized environments
```
Comment (shawkins) [inline crypto/default/src/main/java/org/keycloak/crypto/hash/Argon2PasswordHashProviderFactory.java:29]: Are users going to adjust the memory parameter of Argon2 (for example going up to 46MiB to use a single iteration)? That coupled with running on a node with a lot of cores, like a m5d.24xlarge, with a high or no cpu limit could still be quite a lot of memory.

File: crypto/default/src/main/java/org/keycloak/crypto/hash/Argon2PasswordHashProviderFactory.java [hunk 2]
```diff
@@ -23,6 +24,13 @@ public class Argon2PasswordHashProviderFactory implements PasswordHashProviderFa
...
+     * when a CPU limit is imposed. The throttling would have a negative impact on other concurrent non-hashing activities of Keycloak.
+     */
+    private static final Semaphore CPU_CORE_SEMAPHORE = new Semaphore(Runtime.getRuntime().availableProcessors());
```
Comment (stianst) [inline crypto/default/src/main/java/org/keycloak/crypto/hash/Argon2PasswordHashProviderFactory.java:32]: We should probably make this configurable. That could be useful to limit the concurrent hash operations below the number of processors, if there is a large number of processors available.

**Review by stianst — COMMENTED** [review]
Would be good to reflect this in https://github.com/keycloak/keycloak/blob/main/docs/documentation/server_admin/topics/authentication/password-policies.adoc#password-policy-types, as well as https://github.com/keycloak/keycloak/blob/main/docs/documentation/release_notes/topics/25_0_0.adoc#argon2-password-hashing

File: crypto/default/src/main/java/org/keycloak/crypto/hash/Argon2PasswordHashProviderFactory.java (hunk 1 above)
Comment (stianst) [reply crypto/default/src/main/java/org/keycloak/crypto/hash/Argon2PasswordHashProviderFactory.java:29]: I think we should reflect that in docs; and also make the number of parallell hash operations configurable.

From what I found though increasing memory doesn't decrease execution time (at least not on Java with BouncyCastle implementation), so best practice would probably be to use the min memory recommended by OWASP.

**Comment by stianst** [comment]
LGTM In general, but would be good to make it configurable + reflect this in the docs

**Comment by stianst** [comment]
Somewhat related question should we lower the number of parallell requests being handled by default? 200 seems stupidly large ;)

File: crypto/default/src/main/java/org/keycloak/crypto/hash/Argon2PasswordHashProviderFactory.java [hunk 3]
```diff
@@ -23,6 +24,13 @@ public class Argon2PasswordHashProviderFactory implements PasswordHashProviderFa
     public static final String ITERATIONS_KEY = "iterations";
     public static final String PARALLELISM_KEY = "parallelism";
```
[the remark this answers is not in the document]
Comment (ahus1) [reply crypto/default/src/main/java/org/keycloak/crypto/hash/Argon2PasswordHashProviderFactory.java:25]: I'd opt not to, as 1 is the recommentation of the OAWSP anyway.

File: crypto/default/src/main/java/org/keycloak/crypto/hash/Argon2PasswordHashProviderFactory.java (hunk 2 above)
Comment (ahus1) [reply crypto/default/src/main/java/org/keycloak/crypto/hash/Argon2PasswordHashProviderFactory.java]: This is now configurable, and reflected in the docs.

File: crypto/default/src/main/java/org/keycloak/crypto/hash/Argon2PasswordHashProvider.java [hunk 4]
```diff
@@ -98,19 +101,29 @@ public boolean verify(String rawPassword, PasswordCredentialModel credential) {
...
-        generator.init(parameters);
-
-        byte[] result = new byte[hashLength];
-        generator.generateBytes(rawPassword.toCharArray(), result);
-        return Base64.encodeBytes(result);
+        try {
+            try {
+                cpuCoreSemaphore.acquire();
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
```
Comment (stianst) [inline crypto/default/src/main/java/org/keycloak/crypto/hash/Argon2PasswordHashProvider.java:108]: Isn't the thread already interrupted when an InterruptException is thrown? So is this line needed?

**Review by stianst — APPROVED** [review]

File: crypto/default/src/main/java/org/keycloak/crypto/hash/Argon2PasswordHashProvider.java (hunk 4 above)
Comment (ahus1) [reply crypto/default/src/main/java/org/keycloak/crypto/hash/Argon2PasswordHashProvider.java:108]: An "InterruptedException" is only caught once in a thread. If you don't want to have the next item in the control flow of your thread to also see this interruption as you don't plan to handle the interruption, you'll need to add the flag to the current thread again. 

https://stackoverflow.com/questions/4906799/why-invoke-thread-currentthread-interrupt-in-a-catch-interruptexception-block

**Pull request merged**